### PR TITLE
feat: redesign send rpc transfer flow

### PR DESCRIPTION
**Pull request by alter-eggo**
> Try out Leather build f56dbcd — [Extension build](https://github.com/leather-io/extension/actions/runs/12380601729), [Test report](https://leather-io.github.io/playwright-reports/feat/redesign-send-transfer), [Storybook](https://feat/redesign-send-transfer--65982789c7e2278518f189e7.chromatic.com), [Chromatic](https://www.chromatic.com/library?appId=65982789c7e2278518f189e7&branch=feat/redesign-send-transfer)

redesign rpc send transfer using approver ux

### CODE REVIEW

File: src/app/common/focus-tab.ts
Comment (kyranjamie) [inline src/app/common/focus-tab.ts]: I've refactored this out in my PR, which I'll merge shortly. We shoudln't fire this analytics event directly in the handler for example.

File: src/app/common/transactions/bitcoin/use-generate-bitcoin-tx.ts
Comment (kyranjamie) [inline src/app/common/transactions/bitcoin/use-generate-bitcoin-tx.ts]: Wonder if this will cause broken behaviour elsewhere?

File: src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts
Comment (kyranjamie) [inline src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts]: violates this rule imo https://github.com/ryanmcdermott/clean-code-javascript?tab=readme-ov-file#dont-use-flags-as-function-parameters

Can it be two separate fns? When calling `getFeeForList(x, true)` it isn't clear what true does.

File: src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts
Comment (kyranjamie) [inline src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts]: Shouldn't these presentation/formatting concerns be directly in component?

File: src/app/components/error/form-error.tsx
Comment (kyranjamie) [inline src/app/components/error/form-error.tsx]: this line does nothing as we override in span

File: src/app/components/fees/fee-item-icon.tsx
Comment (kyranjamie) [inline src/app/components/fees/fee-item-icon.tsx]: Correct me if mistaken, but in React this is bad for perf. as it can't do any caching of the JSX and rerenders it every time. 

Better written as:


```tsx
const feeTypeToIconMap: Record<FeeType, ReactNode> = {
  slow: <AnimalSnailIcon variant="small" />,
  standard: <AnimalRabbitIcon variant="small" />,
  fast: <AnimalEagleIcon variant="small" />,
  custom: <AnimalChameleonIcon variant="small" />,
};

export function FeeItemIcon({ feeType }: { feeType: FeeType }) {
  const icon = feeTypeToIconMap[feeType] || null;

  if (!icon) {
    throw new Error('Invalid fee type');
  }

  return <IconWrapper>{icon}</IconWrapper>;
}
```

This example isn't even optimal, as we render all 4 instead of just one, but still more performant than everytime afaik

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (kyranjamie) [inline src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: Better a `useMemo` than a fn. This rerenders on every render, vs on dependency change

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (kyranjamie) [inline src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: Component shouldn't be defined inside another component. Should be defined outside, with props

**Review by kyranjamie — COMMENTED** [review]
Great work @alter-eggo, leaving some comments.

I'll merge my PR now, which has some of the changes to focusTab helper you can use here

File: src/app/common/transactions/bitcoin/use-generate-bitcoin-tx.ts
Comment (alter-eggo) [reply src/app/common/transactions/bitcoin/use-generate-bitcoin-tx.ts]: yeah, I will pass `throwError` arg for now , return null on error is bad here overal

File: src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts
Comment (alter-eggo) [reply src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts]: we will use it in other bitcoin flows as well, so prefer keep in this file, though took it out of fees hook

File: src/app/components/fees/fee-item-icon.tsx
Comment (alter-eggo) [reply src/app/components/fees/fee-item-icon.tsx]: hmm, yeah you are right here, will use your approach

Comment (alter-eggo) [reply src/app/components/fees/fee-item-icon.tsx]: feature is quite simple from perf point of view, so don't think there might be any problems in any approach

File: src/app/common/transactions/bitcoin/use-generate-bitcoin-tx.ts
Comment (pete-watters) [inline src/app/common/transactions/bitcoin/use-generate-bitcoin-tx.ts:27]: Is there an issue for this work? There's none in the commit messages either. 

It would be nice to mention an issue here / in the commit message so we don't forget about this

File: src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts
Comment (pete-watters) [inline src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts:1]: How about splitting up this file and adding some unit tests?

File: src/app/common/transactions/bitcoin/use-generate-bitcoin-tx.ts
Comment (alter-eggo) [reply src/app/common/transactions/bitcoin/use-generate-bitcoin-tx.ts:27]: it would be automatically done when redesign btc send flow to approver ux

File: src/app/components/fees/fee-item.tsx
Comment (pete-watters) [inline src/app/components/fees/fee-item.tsx:38]: Maybe we can add a `FeeButton` wrapper to avoid repeating some stuff e.g. `Add margin compensation when not selected to maintain consistent size`

File: src/app/components/rpc/rpc-header.tsx
Comment (pete-watters) [inline src/app/components/rpc/rpc-header.tsx]: This URL should be a constant

File: src/app/pages/rpc-send-transfer/rpc-send-transfer-choose-fee.tsx
Comment (pete-watters) [inline src/app/pages/rpc-send-transfer/rpc-send-transfer-choose-fee.tsx]: How about taking this out of here to its own file and adding some tests?

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (pete-watters) [inline src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: Should there be two `<SkeletonLoader`s?

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (pete-watters) [inline src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: Maybe you can use a `.layout` file to help split this up between logic and presentation?

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (pete-watters) [inline src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx:89]: This URL should be a constant

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (pete-watters) [inline src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: Did publishing the `preset` package mean this button colour started working properly?

**Review by pete-watters — COMMENTED** [review]
Great work here @alter-eggo 🚀  huge PR! 

I added a few suggestions but looking good to me overall

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: yeah, there are two buttons

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx:89]: agree https://github.com/leather-io/mono/pull/725

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: yeah, it works

File: src/app/components/rpc/rpc-header.tsx
Comment (alter-eggo) [reply src/app/components/rpc/rpc-header.tsx]: https://github.com/leather-io/mono/pull/725

File: src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts
Comment (alter-eggo) [reply src/app/components/bitcoin-fees-list/use-bitcoin-fees-data.ts:1]: yeah, removed all utils to sep file and added tests

File: src/app/components/fees/fee-item.tsx
Comment (alter-eggo) [reply src/app/components/fees/fee-item.tsx:38]: they have quite a lot of differences,, on click action, border color, opacity, don't think there is sense to create wrapper

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: don't understand how to split tbh and why need? what should be in .layout?

File: src/app/pages/rpc-send-transfer/rpc-send-transfer-choose-fee.tsx
Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer-choose-fee.tsx]: moved to utils, added tests

**Review by fbwoolf — APPROVED** [review]
Nice work @alter-eggo! I migrated some of the bitcoin fees code to the mono repo for mobile. I have an issue open to install it back into the extension to make sure it is shared, but that can come later.

File: src/app/components/bitcoin-fees-list/bitcoin-fees.utils.ts
Comment (kyranjamie) [inline src/app/components/bitcoin-fees-list/bitcoin-fees.utils.ts]: Building on earlier comment, why don't we do the `capitalize` in the view too? How we format/present data is a concern of the UI so better lives where it's set to a component.

My thinking is that if I as a dev need to go and make visual changes, they can be mostly scoped to a set of `.layout` components. But in this file, we have a mix of mathematics and formatting, that I can't help but feel should live separately.

e.g. `<style.span>{capitalize(type)}</styled.span>`

Then, the utils file knows nothing about where (`titleLeft`), or how (capitalised), the data needs to be displayed, focusing on accurate data only.

What's your thinking for putting it here, and are there reasons I'm missing why it's beneficial to group it here?

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (kyranjamie) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: There are components with many style props, can't they be in a layout component? 

In `getAddresses` the entire Approver component is in the layout, such that the root RPC compoent can be focused on business logic/event passing behaviours, and not layout and styles.

Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: looking at this file I believe main problem here is with huge approve logic. I will try to move this to sep file.
actually we'll need this in other bitcoin related flows as well

Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: same with selected fee

Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: and prob then will be way easier to move it to layout file. will try

File: src/app/components/fees/fee-item.tsx
Comment (pete-watters) [reply src/app/components/fees/fee-item.tsx:38]: OK 👍 it just seemed like a lot of repeated stuff at first glance, fair enough 👍

File: src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx
Comment (pete-watters) [reply src/app/pages/rpc-send-transfer/rpc-send-transfer.tsx]: Great 👍 so I guess the problem we have there is linking in the `mono` repo between `preset` + `ui`. Thanks for confirming

**Comment by markmhendrickson** [comment]
Looking good! Some QA feedback:

- App favicon isn't showing up in the header
- "Total spend" doesn't include the fee amount
- Guide URL takes user to https://leather.io/guides/connect-dapps instead of a relevant guide for sending BTC via an app
- Do we want to say "Send bitcoin" instead of "Send BTC"? cc @fabric-8 @mica000 
- I'd suggest some additional design QA (by a designer) to double check text styling / sizes, padding, etc.
- Header for edit fee screen has unexpected top and bottom padding in beige

image
image

File: src/app/components/bitcoin-fees-list/bitcoin-fees.utils.ts
Comment (alter-eggo) [reply src/app/components/bitcoin-fees-list/bitcoin-fees.utils.ts]: I tried to do `FeeItem` as dumb as possible, so general idea here was to gen and pass ready to use parts of the fee component, title/caption left/right

yeah, maybe better to `capitalize` in component itself, and so remove titleLeft, I just thought it would be better to keep all these parts in one place.

so in stx I can do the same - get raw fee => then gen ready to use parts of fee component

> Then, the utils file knows nothing about where (titleLeft), or how (capitalised), the data needs to be displayed, focusing on accurate data only.

if not utils so what's the best place to keep it? other way btw - create bitcoin fee wrapper for shared fee item and keep this logic there

**Comment by fabric-8** [comment]
Looking great, only some minor feedback:

CleanShot 2024-12-13 at 15 56 38@2x
- animal icon stroke width are too thin, should generally be same as for the other icons. Maybe they are getting scaled because there's a mismatch of height/width it needs (32x32) vs. actual size after some... padding is applied? If I can help by adjusting the asset lmk, the animation icons are 32x32 vs. 24x24, if turning them into 24x24 fixes this issue I can do it

CleanShot 2024-12-13 at 15 54 43@2x
- Since were are using the `>` to indicate that the user can interact with the item (like for account on top), I think we should also add the `>` to fee settings for consistency cc @mica000 

CleanShot 2024-12-13 at 16 01 53@2x
- Realising that "cancel" in fee settings might get misinterpreted to "cancel the whole approval thing" - Maybe using the word "back" would be better here? cc @markmhendrickson 

- Let's add `border-top: 1px solid ink.border-default` to the footer

**Comment by mica000** [comment]
This is looking great @alter-eggo! Really nice to see it coming together! Bravo 👏

**Comment by mica000** [comment]
Are we covering advanced details as well on this PR? And is this only for BTC or for STX as well?
If so, here are the details we would need to trigger for [Asset transfer BTC](https://www.figma.com/design/FlmF7ropcijOb7iVSDqNn0/%F0%9F%9F%AA-App-Explorations-2024-H2?node-id=735-107715&t=po3Fmhs5V7Ex44B7-4) and [AssetTransferSTX](https://www.figma.com/design/FlmF7ropcijOb7iVSDqNn0/%F0%9F%9F%AA-App-Explorations-2024-H2?node-id=735-110937&t=po3Fmhs5V7Ex44B7-4)

**Comment by mica000** [comment]
@fabric-8 @markmhendrickson Curious about what you think of the prompt, as it shouldn't show the icon of the application since we are already showing this on the request by. However, it is quite nice to have some reassuring of the app you are connecting to:
CleanShot 2024-12-13 at 18 12 23@2x


Here is how it was intended to work and [the respective designs](https://www.figma.com/design/FlmF7ropcijOb7iVSDqNn0/%F0%9F%9F%AA-App-Explorations-2024-H2?node-id=735-107715&t=po3Fmhs5V7Ex44B7-4)
CleanShot 2024-12-13 at 18 10 12@2x

**Comment by mica000** [comment]
@markmhendrickson @fabric-8 We might not have explicitly specified this in the approval documentation, but when connecting to a BTC account, the total balance should be calculated solely for BTC. Including other assets in the calculation could confuse users, especially if they meet the required balance but not in BTC. Thoughts? And if so, should this be in this PR or a later improvment?

**Comment by mica000** [comment]
Account selection: When selecting an account does it make sense to show create account cta? Maybe better to hide it?
CleanShot 2024-12-13 at 17 41 19@2x

**Comment by mica000** [comment]
> Looking great, only some minor feedback:
> 
> CleanShot 2024-12-13 at 15 56 38@2x
> 
> * animal icon stroke width are too thin, should generally be same as for the other icons. Maybe they are getting scaled because there's a mismatch of height/width it needs (32x32) vs. actual size after some... padding is applied? If I can help by adjusting the asset lmk, the animation icons are 32x32 vs. 24x24, if turning them into 24x24 fixes this issue I can do it
> 
> CleanShot 2024-12-13 at 15 54 43@2x
> 
> * Since were are using the `>` to indicate that the user can interact with the item (like for account on top), I think we should also add the `>` to fee settings for consistency cc @mica000
> 
> CleanShot 2024-12-13 at 16 01 53@2x
> 
> * Realising that "cancel" in fee settings might get misinterpreted to "cancel the whole approval thing" - Maybe using the word "back" would be better here? cc @markmhendrickson
> * Let's add `border-top: 1px solid ink.border-default` to the footer

Yes for adding `>` to fee settings for consistency

**Comment by markmhendrickson** [comment]
> Realising that "cancel" in fee settings might get misinterpreted to "cancel the whole approval thing" - Maybe using the word "back" would be better here? cc @markmhendrickson 

Agreed

> Curious about what you think of the prompt, as it shouldn't show the icon of the application since we are already showing this on the request by. However, it is quite nice to have some reassuring of the app you are connecting to:

I do like the idea of including the app icon, though it takes up a lot of space in the current prompt design. It seems perhaps best to just rely on this placement so we don't push the other content below the fold?

image

> We might not have explicitly specified this in the approval documentation, but when connecting to a BTC account, the total balance should be calculated solely for BTC. Including other assets in the calculation could confuse users, especially if they meet the required balance but not in BTC. Thoughts? And if so, should this be in this PR or a later improvment?

It seems best that we either 1) show just a fiat value (for the total account value including all assets, as here in the PR currently) or 2) show the fiat value for the specific asset in question (e.g. BTC value in USD here) **_and_** that asset's literal value (e.g. BTC balance itself here) right below it, as in your designs. By placing these two values together vertically, it'll presumably be clear to the user just which subset of value they're looking at it in this context.

We could add this enhancement (2) later or tackle as part of this PR – both work for me.

**Comment by alter-eggo** [comment]
- removed app icon
- fixed icons
- added chevron in fees item and border in footer actions
- changed `cancel` to `back`
  
Screenshot 2024-12-16 at 15 43 30
Screenshot 2024-12-16 at 15 43 00
Screenshot 2024-12-16 at 15 42 52

@mica000 re: `create new acc` btn - I'd prefer to do it in sep pr since will lead to some refactoring, which I don't want to do here, since pr is already quite big

**Comment by mica000** [comment]
Thanks @alter-eggo!

Icons for fees remains 1 px stroke? 
CleanShot 2024-12-16 at 11 50 34@2x

**Comment by mica000** [comment]
@alter-eggo About the account picker cta, no worries!

In relation to the component on the approval however, lets do the second option suggested by @markmhendrickson in this comment?

> 2) show the fiat value for the specific asset in question (e.g. BTC value in USD here) and that asset's literal value (e.g. BTC balance itself here) right below it, as in your designs. By placing these two values together vertically, it'll presumably be clear to the user just which subset of value they're looking at it in this context. 

In add, @markmhendrickson I believe we can hide one of the address, depending on the network being used.
so it would look more like this:
CleanShot 2024-12-16 at 12 00 09@2x

**Comment by alter-eggo** [comment]
@mica000 done
Screenshot 2024-12-17 at 14 01 40 1

**Comment by mica000** [comment]
@alter-eggo thanks! Looking good!

**Comment by markmhendrickson** [comment]
@alter-eggo it appears this change is still needed given my latest testing:

> Guide URL takes user to https://leather.io/guides/connect-dapps instead of a relevant guide for sending BTC via an app

**Pull request merged**
